tensorboard/webapp/plugins/npmi/views/data_selection/results_download/results_download_container.ts

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Component, ChangeDetectionStrategy, OnDestroy} from '@angular/core';
+import {Store} from '@ngrx/store';
+
+import {combineLatest, Subject} from 'rxjs';
+import {takeUntil, map} from 'rxjs/operators';
+
+import {State} from '../../../../../app_state';
+import {
+  getFlaggedAnnotations,
+  getAnnotationData,
+  getRunToMetrics,
+  getMetricFilters,
+} from '../../../store';
+import {getRunSelection} from '../../../../../core/store/core_selectors';
+import {AnnotationDataListing} from '../../../store/npmi_types';
+import {metricIsNpmiAndNotDiff} from '../../../util/metric_type';
+import {convertToCSVResult} from '../../../util/csv_result';
+
+/** @typehack */ import * as _typeHackRxjs from 'rxjs';
+
+@Component({
+  selector: 'npmi-results-download',
+  template: `
+    <results-download-component
+      [flaggedAnnotations]="flaggedAnnotations$ | async"
+      (onDownloadRequested)="downloadRequested()"
+    ></results-download-component>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ResultsDownloadContainer implements OnDestroy {
+  private ngUnsubscribe = new Subject();
+  private flaggedAnnotations: string[] = [];
+  readonly flaggedAnnotations$ = this.store.select(getFlaggedAnnotations);
+  readonly activeRuns$ = this.store.select(getRunSelection).pipe(
+    map((runSelection) => {
+      if (!runSelection) return [];
+      return Array.from(runSelection.entries())
+        .filter((run) => run[1])
+        .map((run) => run[0]);
+    })
+  );
+  private annotationData: AnnotationDataListing = {};
+  private runs: string[] = [];
+  private metrics: string[] = [];
+
+  constructor(private readonly store: Store<State>) {
+    this.flaggedAnnotations$
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((x) => (this.flaggedAnnotations = x));
+    this.store
+      .select(getAnnotationData)
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((x) => (this.annotationData = x));
+    this.activeRuns$
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((x) => (this.runs = x));
```

Comment:
Hm, this is an interesting pattern.  Admittedly, I don't know whether it's a good/bad practice to subscribe manually and set component properties like this.

In many other containers, a custom `createSelector` manipulated the data into the form we need, and passes it via async pipe to a component.  Reasoning about when container/component properties update is simpler because only no local property state in the container needs to be kept.

Wdyt about replacing all the observable subscriptions with
```
annotationData$ = this.store.select(getAnnotationData
flaggedAnnotations$ = this.store.select(getFlaggedAnnotations
metrics$ = createSelector(getRunToMetrics, getMetricFilters, getRunSelection
    (runToMetrics, metricFilters, runSelection, annotationData, flaggedAnnotations) => {
        // all the logic to produce the metrics 
    }
```
then async-pipe'ing these into the component, and moving the entire `downloadRequested()` from container into component?

There would be no local state in this container to maintain, and no need to ngUnsubscribe, since we'd be using async pipes.

Reply:
Sure, I can do that. Just thought that the container should not have such methods as it should not assume anything about the state. And I guess I would implement it like this as I think we dont need a `createSelector` here:
```
readonly metrics$ = combineLatest([
      this.store.select(getRunToMetrics),
      this.activeRuns$,
      this.store.select(getMetricFilters)
    ])
      .pipe(
        map(([runToMetrics, activeRuns, metricFilters]) => {
          let metrics = Object.keys(metricFilters);
          for (const run of activeRuns) {
            if (runToMetrics[run]) {
              metrics = metrics.concat(
                runToMetrics[run].filter((key) => metricIsNpmiAndNotDiff(key))
              );
            }
          }
          metrics = [...new Set(metrics)];
          return metrics;
        })
      )
```

Reply:
As per offline discussion: combineLatest instead of createSelector sounds good.  I'm more concerned about replacing the `.subscribe(callback)`s.  L108 - L114 can move into the component without giving it much business logic.